File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PIDS/PID.java
package org.firstinspires.ftc.teamcode.PIDS;

import com.qualcomm.robotcore.util.ElapsedTime;

import java.util.ArrayList;

public class PID {
    private double kp, ki, kd;
    private double totalError, lastError, lastTime;
    private ArrayList<Double> errors;

    public PID(double kp, double ki, double kd) {
        this.kp = kp;
        this.ki = ki;
        this.kd = kd;
        lastTime = 0;
        errors = new ArrayList<Double>();
    }

    public double getP(double error) {
        return kp * error;
    }

    public double getI(double error) {
        if (Math.abs(error) < 1)
            totalError = 0;
        return ki * totalError;
    }

    public double getD(double currentError, ElapsedTime runtime) {
        return kd * (currentError - lastError)/(runtime.time()-lastTime);
    }

    public boolean checkCorrection(double error){

        if(Math.abs(errors.get(errors.size()-1)) < Math.abs(error))
            return true;
        else
            return false;

    }

    public double getCorrection(double error, ElapsedTime runtime) {
       /* if(errors.size() != 0) {
            totalError += error;
            double potentialError = getP(error) + getI(error) + getD(error, runtime);
            if (Math.abs(error) < 0.001 || (checkCorrection(potentialError))) {
                totalError -= error;
                return 0;
            }
            totalError -= error;
        }*/

        if (Math.abs(error) < 0.001) {
            return 0;
        }

        totalError += error;

        double output = getP(error) + getI(error) + getD(error, runtime);

        lastError = error;
        errors.add(output);
        lastTime = runtime.time();

        return output;
    }

    public ArrayList<Double> aman(){
        return errors;
    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/pids/PID.kt
package org.firstinspires.ftc.teamcode.pids

import com.qualcomm.robotcore.util.ElapsedTime
import kotlin.math.abs

open class PID(protected var kP: Double, protected val kI: Double, protected var kD: Double) {
    protected var totalError = 0.0
    protected var lastError = 0.0
    protected var lastTime = 0.0

    protected val _errors = mutableListOf<Double>()
    val errors: List<Double>
        get() = _errors

    open fun p(error: Double): Double {
        return kP * error
    }

    open fun i(error: Double): Double {
        if (abs(error) < 1) {
            totalError = 0.0
        }
        return kI * totalError
    }

    open fun d(currentError: Double, runtime: ElapsedTime): Double {
        return kD * (currentError - lastError) / (runtime.time() - lastTime)
    }

    open fun checkCorrection(error: Double) = abs(_errors.last()) < abs(error)

    open fun getCorrection(error: Double, runtime: ElapsedTime): Double {
        if (abs(error) < .001) {
            return 0.0
        }

        totalError += error
        lastError = error
        lastTime = runtime.time()

        return p(error) + i(error) + d(error, runtime).also {
            _errors.add(it)
        }
    }
}